### ISSUE TITLE
Support for the "each" iterator in YAML

### ISSUE THREAD

**Issue by scottyw** [lyraproj/lyra#209]
**Is your feature request related to a problem? Please describe.**
We support iterators in YAML including "times" and "range". This ticket captures implementing support for "each" as well.

**Describe the solution you'd like**
Here's an example workflow:
```yaml
aws:
  typespace: TerraformAws
  input:
    subnet_cidr_blocks:
      type: Array
      lookup: subnet_cidr_blocks
  output: subnets
  activities:

    aws_vpc:
      # type is implicit and is derived from the
      # activity name, in contrast to subnets below
      output: aws_vpc_id
      state:
        cidr_block: 192.168.0.0/16
        instance_tenancy: default

    aws_subnet:
      type:  TerraformAws::aws_subnet
      output:
         key: aws_subnet_id
         value: [aws_subnet_id]
      iteration:
         name:     subnets
         function: each
         over:     subnet_cidr_blocks
         vars:     cidr_block
      state:
        vpc_id: $aws_vpc_id
        cidr_block: $cidr_block
        tags:
          created_by: lyra
```
Input data would be something like:
```yaml
subnet_cidr_blocks: [ 192.168.1.0/24, 192.168.2.0/24, 192.168.3.0/24 ]
```

**Comment by jdwelch** [comment]
Not a fan :(

**Comment by thallgren** [comment]
@scottyw the interpolated string `"Subnet covering $cidr_block"` is not going to work. The only way it could be made to work is to interpolate all values regardless of if are quoted or not and regardless of what quotes that are used. The YAML parser strips all such information off. Our evaluator never sees it. The rule today is that a value must match an input variable exactly (quoted or not), only then will it be substituted for the input variable. Should we rethink this?

@jdwelch are you opposed to iterators in YAML as such, or just the `each` variant?

**Comment by scottyw** [comment]
@thallgren I removed that line from the example. I think the interpolation isn't really important.

Regarding iterators in yaml, it's something that we've seen spark a lot of strong feelings. My intention with this ticket was only really to say that we document support for `each` but don't actually support it. The bigger question, asking whether we should do it at all, is probably the important one. I wouldn't mind if we dropped it in the interests of keeping the yaml experience simple-but-limited and anticipate users that need more (including iteration) will move to TS or some other supported programming language.

@jdwelch ^^
